### DISCUSSION TITLE
Can we disable list all records but enable only get by id?

### DISCUSSION THREAD

**Discussion by Albert-Gao** [supabase/supabase#15366] [Questions] [answered]
For example, 
for the profiles table, I do not want to enable the user to list all profiles, 
but a user should be able to get a profile by its id, 

For example, user A can query user B, user C or any arbitrary user by using an ID field, 
but a user can not do a SELECT * without specifying the id, is it possible?

is it possible without creating an API?  
How to do it plz? :)

I think I can do it via creating an RLS which disables all select, and provide an API to getUserByID(), would be great if I can use an RLS for this :) 

Thanks. :)

**Comment by GaryAustin1** [comment, marked as answer]
You need to lock the table select RLS with false.  And then use an rpc call to a security definer function to read the row by provided id.
The security definer function will bypass the RLS to get you want.

Edit.  I see you edited as I was typing.   RLS can't do this on its own.

**Reply by Albert-Gao** [reply]
Thanks for the confirmation, what do you mean by calling a security definer, please? :D
I am currently just calling from admin SDK and everything works great, is admin SDK a security definer?
Thanks :)

**Reply by GaryAustin1** [reply]
If you use service_role then that will bypass RLS.  Postgres functions can also be run as "security definer" or "security invoker".  Invoker runs as the user making the call to the function.  So if a user it must meet RLS.   This is the default.   Definer runs as the owner or creator of the function so if created with the SQL editor (or with the UI) it will bypass RLS.  The UI defaults to invoker.  So if you are calling from a server and using service_role, then you are fine.

**Reply by Albert-Gao** [reply]
> If you use service_role then that will bypass RLS.  Postgres functions can also be run as "security definer" or "security invoker".  Invoker runs as the user making the call to the function.  So if a user it must meet RLS.   This is the default.   Definer runs as the owner or creator of the function so if created with the SQL editor (or with the UI) it will bypass RLS.  The UI defaults to invoker.  So if you are calling from a server and using service_role, then you are fine.

Yes,  the admin function in invoked from server side for preventing leaking the service role key.

Thanks for your time and such detailed explanation. I think it means:

- caller of a DB function is security invoker
- creator/owner of a DB function is security definer
- security definer bypasses all the RLS check

I just have one more question, does it imply that only an DB admin can add a DB function? (Which explains why the bypassing works). Which means, a normal DB user can not create a DB function? 

Sorry, I try to understand from a app engineer point of view, not quite a DB guy 😝
